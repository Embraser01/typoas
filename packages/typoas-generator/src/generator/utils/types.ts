import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { factory, TypeNode, SyntaxKind } from 'typescript';
import { Context } from '../../context';
import {
  hasUnsupportedIdentifierChar,
  sanitizeTypeIdentifier,
} from './operation-name';

const ENUM_SUPPORTED_TYPE = ['integer', 'number', 'string', 'null', 'boolean'];

export function createTypeFromSchema(
  schemaOrRef: SchemaObject | ReferenceObject | undefined,
  ctx: Context,
): TypeNode {
  let node: TypeNode = factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);

  if (!schemaOrRef) {
    return node;
  }
  if (schemaOrRef.$ref) {
    const ref = ctx.resolveReference('schemas', schemaOrRef.$ref);
    if (!ref) {
      throw new Error(`$ref '${schemaOrRef.$ref}' wasn't found`);
    }
    node = factory.createTypeReferenceNode(
      factory.createIdentifier(sanitizeTypeIdentifier(ref.name)),
    );
  }

  const schema: SchemaObject = schemaOrRef;
  if (schema.enum && schema.type && ENUM_SUPPORTED_TYPE.includes(schema.type)) {
    node = factory.createUnionTypeNode(
      schema.enum
        .map((e) => {
          if (e === null) return factory.createNull();
          if (typeof e === 'boolean') {
            return e ? factory.createTrue() : factory.createFalse();
          }
          if (typeof e === 'number') {
            return factory.createNumericLiteral(e);
          }
          return factory.createStringLiteral(e, true);
        })
        .map((e) => factory.createLiteralTypeNode(e)),
    );
  } else if (schema.oneOf && schema.oneOf.length) {
    node = factory.createUnionTypeNode(
      schema.oneOf.map((s) => createTypeFromSchema(s, ctx)),
    );
  } else if (schema.anyOf && schema.anyOf.length) {
    node = factory.createUnionTypeNode(
      schema.anyOf.map((s) => createTypeFromSchema(s, ctx)),
    );
  } else {
    switch (schema.type) {
      case 'integer':
      case 'number':
        node = factory.createKeywordTypeNode(SyntaxKind.NumberKeyword);
        break;
      case 'string':
        switch (schema.format) {
          case 'date':
          case 'date-time':
            node = factory.createTypeReferenceNode(
              factory.createIdentifier('Date'),
            );
            break;
          default:
            node = factory.createKeywordTypeNode(SyntaxKind.StringKeyword);
            break;
        }
        break;
      case 'boolean':
        node = factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword);
        break;
      case 'object': {
        const hasProperties =
          schema.properties && Object.keys(schema.properties).length;

        if (hasProperties) {
          node = factory.createTypeLiteralNode(
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            Object.entries(schema.properties!).map(([key, val]) =>
              factory.createPropertySignature(
                undefined,
                hasUnsupportedIdentifierChar(key)
                  ? factory.createStringLiteral(key, true)
                  : factory.createIdentifier(key),
                schema.required?.includes(key)
                  ? undefined
                  : factory.createToken(SyntaxKind.QuestionToken),
                createTypeFromSchema(val, ctx),
              ),
            ),
          );
        }
        if (schema.additionalProperties) {
          const additionalPropsNode = factory.createTypeLiteralNode([
            factory.createIndexSignature(
              undefined,
              undefined,
              [
                factory.createParameterDeclaration(
                  undefined,
                  undefined,
                  undefined,
                  factory.createIdentifier('key'),
                  undefined,
                  factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                  undefined,
                ),
              ],
              schema.additionalProperties === true
                ? factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)
                : createTypeFromSchema(schema.additionalProperties, ctx),
            ),
          ]);

          node = hasProperties
            ? factory.createIntersectionTypeNode([node, additionalPropsNode])
            : additionalPropsNode;
        }
        break;
      }
      case 'null':
        node = factory.createLiteralTypeNode(factory.createNull());
        break;
      case 'array':
        node = factory.createArrayTypeNode(
          createTypeFromSchema(schema.items, ctx),
        );
        break;
    }
  }

  if (schema.nullable) {
    node = factory.createUnionTypeNode([
      node,
      factory.createLiteralTypeNode(factory.createNull()),
    ]);
  }
  return node;
}
