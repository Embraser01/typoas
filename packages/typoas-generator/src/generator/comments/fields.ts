import {
  addSyntheticLeadingComment,
  createPrinter,
  factory,
  Node,
  SyntaxKind,
  EmitHint,
  NodeFlags,
} from 'typescript';

const jsDocPrinter = createPrinter();

export function addJSDocToNode(node: Node): void {
  const jsDoc = factory.createJSDocComment('Test', []);

  const src = factory.createSourceFile(
    [],
    factory.createToken(SyntaxKind.EndOfFileToken),
    NodeFlags.Const,
  );
  const output = jsDocPrinter
    .printNode(EmitHint.Unspecified, jsDoc, src)
    .trim()
    .replace(/^\/\*|\*\/$/g, '');

  addSyntheticLeadingComment(
    node,
    SyntaxKind.MultiLineCommentTrivia,
    output,
    true,
  );
}
