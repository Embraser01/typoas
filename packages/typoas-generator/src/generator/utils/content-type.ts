import {
  ContentObject,
  ReferenceObject,
  SchemaObject,
} from 'openapi3-ts/oas31';
import { Context } from '../../context';

// This key is a simple hack for the createTypeFromSchema function
// for it to generate a Blob type that does not exist by default in json schema.
export const TYPOAS_BLOB_TYPE_KEY = 'x-typoas-blob-type';

export function getContentTypeSchema(
  content: ContentObject | undefined,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  ctx: Context,
): ReferenceObject | SchemaObject | undefined {
  if (!content) {
    return undefined;
  }
  if (content['application/json']) {
    return content['application/json'].schema;
  }

  if (content['text/plain']) {
    return { type: 'string' };
  }

  if (content['application/octet-stream']) {
    return { [TYPOAS_BLOB_TYPE_KEY]: true };
  }
}
