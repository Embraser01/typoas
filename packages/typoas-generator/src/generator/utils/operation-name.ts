function lowerCase(str: string) {
  return str.toLowerCase();
}

function upperCase(str: string) {
  return str.toUpperCase();
}

/**
 * Convert string to camelCase text.
 */
function camelCase(str: string): string {
  return str
    .replace(/[-./]/g, ' ') //convert all hyphens to spaces
    .replace(/\s[a-z]/g, upperCase) //convert first char of each word to UPPERCASE
    .replace(/\s+/g, '') //remove spaces
    .replace(/^[A-Z]/g, lowerCase); //convert first char to lowercase
}

function pascalCase(str: string): string {
  return camelCase(str).replace(/^[a-z]/, upperCase);
}

export function hasUnsupportedIdentifierChar(key: string): boolean {
  return /[-/.+@]/.test(key);
}

export function sanitizeOperationIdName(op: string): string {
  return camelCase(op);
}

export function sanitizeTypeIdentifier(type: string): string {
  return pascalCase(type);
}
