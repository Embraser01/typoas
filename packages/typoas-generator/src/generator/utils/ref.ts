import { Expression, factory, TypeNode } from 'typescript';

export const IMPORT_RUNTIME = 'r';

export enum ExportedRef {
  HttpMethod = 'HttpMethod',
  BaseServerConfiguration = 'BaseServerConfiguration',
  HttpLibrary = 'HttpLibrary',
  RefResolver = 'RefResolver',
  IsomorphicFetchHttpLibrary = 'IsomorphicFetchHttpLibrary',
  SchemaRefResolver = 'SchemaRefResolver',
  ApiException = 'ApiException',
  AuthProvider = 'AuthProvider',
  BasicAuthConfig = 'BasicAuthConfig',
  BearerAuthConfig = 'BearerAuthConfig',
  BaseFlowConfig = 'BaseFlowConfig',
  SecurityAuthentication = 'SecurityAuthentication',
  ApiKeySecurityAuthentication = 'ApiKeySecurityAuthentication',
  HttpSecurityAuthentication = 'HttpSecurityAuthentication',
  OAuth2SecurityAuthentication = 'OAuth2SecurityAuthentication',
  applyTemplating = 'applyTemplating',
  applyTransforms = 'applyTransforms',
  isCodeInRange = 'isCodeInRange',
  serializeParameter = 'serializeParameter',
}

export function createRuntimeRefProperty(exportedRef: ExportedRef): Expression {
  return factory.createPropertyAccessExpression(
    factory.createIdentifier(IMPORT_RUNTIME),
    exportedRef,
  );
}

export function createRuntimeRefType(
  exportedRef: ExportedRef,
  typeArguments?: readonly TypeNode[],
): TypeNode {
  return factory.createTypeReferenceNode(
    factory.createQualifiedName(
      factory.createIdentifier(IMPORT_RUNTIME),
      factory.createIdentifier(exportedRef),
    ),
    typeArguments,
  );
}

export function createJSONParseWrapper(data: unknown): Expression {
  return factory.createCallExpression(
    factory.createPropertyAccessExpression(
      factory.createIdentifier('JSON'),
      'parse',
    ),
    undefined,
    [factory.createStringLiteral(JSON.stringify(data), true)],
  );
}
