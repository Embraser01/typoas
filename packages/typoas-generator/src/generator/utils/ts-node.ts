import {
  createPrinter,
  EmitHint,
  factory,
  NewLineKind,
  Node,
  NodeFlags,
  SyntaxKind,
} from 'typescript';

export const getStringFromNode = (node: Node): string => {
  const printer = createPrinter({ newLine: NewLineKind.LineFeed });
  const resultFile = factory.createSourceFile(
    [],
    factory.createToken(SyntaxKind.EndOfFileToken),
    NodeFlags.Const,
  );
  return printer.printNode(EmitHint.Unspecified, node, resultFile);
};
