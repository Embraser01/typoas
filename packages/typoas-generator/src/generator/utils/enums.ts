import { SchemaObject } from 'openapi3-ts';
import { factory, EnumMember } from 'typescript';
import { screamingSnakeCase } from './operation-name';

export function canConvertSchemaToEnum(schema: SchemaObject): boolean {
  return (
    schema.type === 'string' &&
    schema.enum !== undefined &&
    schema.enum.every((e) => ![null, '', undefined].includes(e)) &&
    !schema.nullable
  );
}

export function createEnumMembersFromSchema(
  schema: SchemaObject,
): EnumMember[] {
  if (canConvertSchemaToEnum(schema)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return schema.enum!.map((e) => {
      return factory.createEnumMember(
        factory.createIdentifier(screamingSnakeCase(e)),
        factory.createStringLiteral(e, true),
      );
    });
  }
  throw new Error('Cannot create enum members from this schema');
}
