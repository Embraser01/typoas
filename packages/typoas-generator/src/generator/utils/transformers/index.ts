import {
  ArrayLiteralExpression,
  Expression,
  factory,
  ObjectLiteralElementLike,
  ObjectLiteralExpression,
} from 'typescript';
import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { Context } from '../../../context';
import { isTransformableLeafDate } from './date';
import { findLeafTransforms, TransformField } from './leaf-transformer-base';
import { createRuntimeRefProperty, ExportedRef } from '../ref';

export function createSchemaTransforms(
  schema: SchemaObject | ReferenceObject,
  ctx: Context,
): ObjectLiteralExpression | null {
  const transforms: ObjectLiteralElementLike[] = [];

  // Date transform
  const rawTransforms = findLeafTransforms(schema, isTransformableLeafDate);
  if (rawTransforms.length) {
    transforms.push(
      factory.createPropertyAssignment(
        'date',
        createFromRawTransforms(rawTransforms, ctx),
      ),
    );
  }

  if (!transforms.length) {
    return null;
  }
  return factory.createObjectLiteralExpression(transforms);
}

function createFromRawTransforms(
  rawTransforms: TransformField[],
  ctx: Context,
): ArrayLiteralExpression {
  return factory.createArrayLiteralExpression(
    rawTransforms.map((transformFields) =>
      factory.createArrayLiteralExpression(
        transformFields.map(([type, data]) => {
          const items: Expression[] = [
            factory.createPropertyAccessExpression(
              createRuntimeRefProperty(ExportedRef.TransformType),
              factory.createIdentifier(type.toUpperCase()),
            ),
          ];

          if (typeof data === 'string') {
            items.push(factory.createStringLiteral(data));
          } else if (Array.isArray(data)) {
            items.push(createFromRawTransforms(data, ctx));
          }

          return factory.createArrayLiteralExpression(items);
        }),
      ),
    ),
  );
}
