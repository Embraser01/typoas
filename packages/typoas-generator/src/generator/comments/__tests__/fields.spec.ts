import { describe, expect, it } from '@jest/globals';
import { factory, SyntaxKind } from 'typescript';
import { getStringFromNode } from '../../utils/ts-node';
import { addJSDocToNode } from '../fields';

describe('jsdoc on fields', () => {
  it('should add a simple jsdoc comment to a node', () => {
    const node = factory.createTypeAliasDeclaration(
      undefined,
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      factory.createIdentifier('MyType'),
      undefined,
      factory.createKeywordTypeNode(SyntaxKind.AnyKeyword),
    );

    addJSDocToNode(node);

    expect(getStringFromNode(node)).toMatchSnapshot();
  });
});
