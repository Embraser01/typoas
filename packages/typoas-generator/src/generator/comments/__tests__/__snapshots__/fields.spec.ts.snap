// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsdoc on fields should add a simple jsdoc comment to a node 1`] = `
"/**
 * Test
 */
export type MyType = any;"
`;
